tests/components/zwave_js/test_api.py

```diff
@@ -4674,6 +4681,11 @@ async def test_hard_reset_controller(
     assert len(client.async_send_command.call_args_list) == 1
     assert client.async_send_command.call_args[0][0] == {"command": "driver.hard_reset"}
 
+    async_dispatcher_send(hass, EVENT_DEVICE_ADDED_TO_REGISTRY, device)
```

Comment:
Can't we add a node instead do that this message is dispatched by the integration?

Reply:
really what we should do is restart the config entry. When we hard reset, the driver restarts, and then we get the controller state in the state dump when we `start_listening`. So reloading the config entry mimicks this behavior, just without the change of the home ID. Thoughts?

Reply:
Here are the logs starting from reset:
```
2023-10-19 09:00:53.008 DEBUG (MainThread) [zwave_js_server] Publishing message:
{'command': 'driver.hard_reset',
 'messageId': '880b32b033494dfa8930e730ae2082cd'}

2023-10-19 09:00:53.015 DEBUG (MainThread) [zwave_js_server] Received message:
WSMessage(type=<WSMsgType.TEXT: 1>, data='{"type":"result","success":true,"messageId":"880b32b033494dfa8930e730ae2082cd","result":{}}', extra='')

2023-10-19 09:00:53.017 DEBUG (MainThread) [zwave_js_server] Listen completed. Cleaning up
2023-10-19 09:00:53.017 INFO (MainThread) [homeassistant.components.zwave_js] Disconnected from server. Reloading integration
2023-10-19 09:00:53.020 DEBUG (MainThread) [zwave_js_server] Trying to connect
2023-10-19 09:00:58.354 DEBUG (MainThread) [zwave_js_server] Trying to connect
2023-10-19 09:00:58.366 DEBUG (MainThread) [zwave_js_server] Received message:
WSMessage(type=<WSMsgType.TEXT: 1>, data='{"type":"version","driverVersion":"12.2.0","serverVersion":"1.32.2","homeId":4087487182,"minSchemaVersion":0,"maxSchemaVersion":32}', extra='')

2023-10-19 09:00:58.366 INFO (MainThread) [zwave_js_server] Connected to Home 4087487182 (Server 1.32.2, Driver 12.2.0, Using Schema 32)
2023-10-19 09:00:58.366 INFO (MainThread) [homeassistant.components.zwave_js] Connected to Zwave JS Server
2023-10-19 09:00:58.371 DEBUG (MainThread) [zwave_js_server] Publishing message:
{'additionalUserAgentComponents': {'HomeAssistant': '2023.11.0.dev0',
                                   'zwave-js-server-python': '0.52.1'},
 'command': 'initialize',
 'messageId': 'initialize',
 'schemaVersion': 32}

2023-10-19 09:00:58.378 DEBUG (MainThread) [zwave_js_server] Received message:
WSMessage(type=<WSMsgType.TEXT: 1>, data='{"type":"result","success":true,"messageId":"initialize","result":{}}', extra='')

2023-10-19 09:00:58.378 DEBUG (MainThread) [zwave_js_server] Publishing message:
{'command': 'driver.get_log_config', 'messageId': 'get-initial-log-config'}

2023-10-19 09:00:58.385 DEBUG (MainThread) [zwave_js_server] Received message:
WSMessage(type=<WSMsgType.TEXT: 1>, data='{"type":"result","success":true,"messageId":"get-initial-log-config","result":{"config":{"enabled":true,"level":"debug","logToFile":false,"maxFiles":7,"filename":"/usr/src/node_modules/@zwave-js/server/dist/bin/zwavejs_%DATE%.log","forceConsole":true}}}', extra='')

2023-10-19 09:00:58.385 DEBUG (MainThread) [zwave_js_server] Publishing message:
{'command': 'start_listening', 'messageId': 'start-listening'}

2023-10-19 09:00:58.399 DEBUG (MainThread) [zwave_js_server] Received message:
WSMessage(type=<WSMsgType.TEXT: 1>, data='{"type":"result","success":true,"messageId":"start-listening","result":{"state":{"driver":{"logConfig":{"enabled":true,"level":"debug","logToFile":false,"maxFiles":7,"filename":"/usr/src/node_modules/@zwave-js/server/dist/bin/zwavejs_%DATE%.log","forceConsole":true},"statisticsEnabled":false},"controller":{"type":7,"homeId":4087487182,"ownNodeId":1,"isUsingHomeIdFromOtherNetwork":false,"isSISPresent":false,"wasRealPrimary":true,"manufacturerId":0,"productType":4,"productId":4,"supportedFunctionTypes":[2,3,5,6,7,8,9,10,11,16,18,19,20,21,22,28,32,33,34,35,36,39,40,41,42,44,46,55,56,57,58,59,60,63,65,66,68,69,70,71,72,73,74,75,77,79,80,81,83,84,85,86,87,88,94,95,96,97,98,99,101,128,144,146,147,152,160,162,164,165,166,168,169,171,189,190,191,210,211,212,214,215,216,217,218,219,220,221,222,223,231,232],"sucNodeId":1,"supportsTimers":false,"statistics":{"messagesTX":2845,"messagesRX":2835,"messagesDroppedRX":0,"NAK":0,"CAN":0,"timeoutACK":0,"timeoutResponse":0,"timeoutCallback":0,"messagesDroppedTX":0,"backgroundRSSI":{"channel0":{"current":-90,"average":-87},"channel1":{"current":-90,"average":-87},"channel2":{"current":-90,"average":-87},"timestamp":1697720429395}},"inclusionState":0,"sdkVersion":"7.18.3","firmwareVersion":"7.18","isPrimary":true,"isSUC":false,"nodeType":0,"rfRegion":1,"status":0,"isRebuildingRoutes":false},"nodes":[{"nodeId":1,"index":0,"status":4,"ready":true,"isListening":true,"isRouting":true,"manufacturerId":0,"productId":4,"productType":4,"firmwareVersion":"7.18.3","deviceConfig":{"filename":"/data/db/devices/0x0000/700_series_controller.json","isEmbedded":true,"manufacturer":"Silicon Labs","manufacturerId":0,"label":"700 Series","description":"USB Controller","devices":[{"productType":4,"productId":4}],"firmwareVersion":{"min":"0.0","max":"255.255"},"preferred":false,"metadata":{"comments":[]}},"label":"700 Series","interviewAttempts":1,"endpoints":[{"nodeId":1,"index":0,"deviceClass":{"basic":{"key":2,"label":"Static Controller"},"generic":{"key":2,"label":"Static Controller"},"specific":{"key":7,"label":"Gateway"},"mandatorySupportedCCs":[114,134],"mandatoryControlledCCs":[32,96]},"commandClasses":[{"id":114,"name":"Manufacturer Specific","version":2,"isSecure":false},{"id":134,"name":"Version","version":3,"isSecure":false}]}],"values":[{"endpoint":0,"commandClass":114,"commandClassName":"Manufacturer Specific","property":"manufacturerId","propertyName":"manufacturerId","ccVersion":0,"metadata":{"type":"number","readable":true,"writeable":false,"label":"Manufacturer ID","min":0,"max":65535,"stateful":true,"secret":false},"value":0},{"endpoint":0,"commandClass":114,"commandClassName":"Manufacturer Specific","property":"productType","propertyName":"productType","ccVersion":0,"metadata":{"type":"number","readable":true,"writeable":false,"label":"Product type","min":0,"max":65535,"stateful":true,"secret":false},"value":4},{"endpoint":0,"commandClass":114,"commandClassName":"Manufacturer Specific","property":"productId","propertyName":"productId","ccVersion":0,"metadata":{"type":"number","readable":true,"writeable":false,"label":"Product ID","min":0,"max":65535,"stateful":true,"secret":false},"value":4},{"endpoint":0,"commandClass":134,"commandClassName":"Version","property":"firmwareVersions","propertyName":"firmwareVersions","ccVersion":0,"metadata":{"type":"string[]","readable":true,"writeable":false,"label":"Z-Wave chip firmware versions","stateful":true,"secret":false},"value":["7.18"]},{"endpoint":0,"commandClass":134,"commandClassName":"Version","property":"zWaveProtocolVersion","propertyName":"zWaveProtocolVersion","ccVersion":0,"metadata":{"type":"string","readable":true,"writeable":false,"label":"Z-Wave protocol version","stateful":true,"secret":false},"value":"7.18.3"},{"endpoint":0,"commandClass":134,"commandClassName":"Version","property":"sdkVersion","propertyName":"sdkVersion","ccVersion":0,"metadata":{"type":"string","readable":true,"writeable":false,"label":"SDK version","stateful":true,"secret":false},"value":"7.18.3"}],"isFrequentListening":false,"maxDataRate":100000,"supportedDataRates":[40000,100000],"protocolVersion":3,"supportsBeaming":true,"supportsSecurity":false,"nodeType":0,"deviceClass":{"basic":{"key":2,"label":"Static Controller"},"generic":{"key":2,"label":"Static Controller"},"specific":{"key":7,"label":"Gateway"},"mandatorySupportedCCs":[114,134],"mandatoryControlledCCs":[32,96]},"interviewStage":"Complete","deviceDatabaseUrl":"https://devices.zwave-js.io/?jumpTo=0x0000:0x0004:0x0004:7.18.3","statistics":{"commandsTX":0,"commandsRX":0,"commandsDroppedRX":0,"commandsDroppedTX":0,"timeoutResponse":0},"isControllerNode":true,"keepAwake":false}]}}}', extra='')

2023-10-19 09:00:58.401 INFO (MainThread) [zwave_js_server] Z-Wave JS initialized. 1 nodes
2023-10-19 09:00:58.401 INFO (MainThread) [homeassistant.components.zwave_js] Connection to Zwave JS Server initialized
2023-10-19 09:00:58.401 INFO (MainThread) [homeassistant.components.zwave_js] Enabling zwave-js-server logging
2023-10-19 09:00:58.402 DEBUG (MainThread) [zwave_js_server] Publishing message:
{'command': 'start_listening_logs',
 'messageId': 'cb657473b6064021b594deace635804d'}

2023-10-19 09:00:58.411 DEBUG (MainThread) [zwave_js_server] Received message:
WSMessage(type=<WSMsgType.TEXT: 1>, data='{"type":"result","success":true,"messageId":"cb657473b6064021b594deace635804d","result":{}}', extra='')

2023-10-19 09:00:58.411 INFO (MainThread) [homeassistant.components.zwave_js] Zwave-js-server logging is enabled
2023-10-19 09:00:58.416 INFO (MainThread) [homeassistant.components.sensor] Setting up sensor.zwave_js
2023-10-19 09:00:58.422 DEBUG (MainThread) [homeassistant.components.zwave_js] Node added: 1
2023-10-19 09:00:58.422 DEBUG (MainThread) [homeassistant.components.zwave_js] Processing node Node(node_id=1)
```

Reply:
Can we mimic closing the connection in the client, so that the integration reloads the config entry itself?

I'm thinking we could replace this `asyncio.Event` with an `asyncio.Future`.

https://github.com/home-assistant/core/blob/6baa8082d572de84fccb5b5ee3697882fa25fa05/tests/components/zwave_js/conftest.py#L684

Then we can set a result when we want to stop waiting for the future, or eg raise an exception for other cases. If we define and return the future from a separate fixture we can access it in any test.

Reply:
One problem could be that we'd need to create a new future when reloading the config entry.

Another way could be to just extract the `asyncio.Event` to a fixture so that we can set and clear the event when needed.